pkg/auth/authmap_gc.go

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/cilium/cilium/pkg/identity"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/node/addressing"
+)
+
+type authMapGarbageCollector struct {
+	authmap authMap
+	ipCache ipCache
+
+	discoveredCiliumNodeIDs    map[uint16]struct{}
+	discoveredCiliumIdentities map[identity.NumericIdentity]struct{}
+}
+
+func newAuthMapGC(authmap authMap, ipCache ipCache) *authMapGarbageCollector {
+	return &authMapGarbageCollector{
+		authmap: authmap,
+		ipCache: ipCache,
+		discoveredCiliumNodeIDs: map[uint16]struct{}{
+			0: {}, // Local node 0 is always available
+		},
+		discoveredCiliumIdentities: map[identity.NumericIdentity]struct{}{},
+	}
+}
+
+func (r *authMapGarbageCollector) HandleCiliumNodeEvent(_ context.Context, e resource.Event[*ciliumv2.CiliumNode]) error {
+	var err error
+	defer e.Done(err)
```

Comment:
This doesn't work. the `err` is evaluated eagerly here and will always be nil. You need to do `func() { e.Done(err) }`.
https://go.dev/play/p/xbVK-puk5cD.

I would do `(err error)` return value so we'll capture the context too (e.g. the "failed to cleanup missing codes"). Consider adding unit tests to verify that the right thing happens. `type Event struct` makes this pretty easy as `Done` is just a field.

@dylandreimerink double-check that this isn't happening in your code as well and that retry handling isn't broken because of this.

Reply:
@joamaki :facepalm: of course, thanks. 

fixed it and added unit-tests which are checking the returned and also the reported error (via event done)